cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/graph.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/thrust_tuple_utils.hpp>
+#include <cugraph/vertex_partition_device_view.cuh>
+
+#include <sampling/detail/sampling_utils.hpp>
+
+#include <raft/core/handle.hpp>
+
+#include <thrust/optional.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+#include <rmm/device_uvector.hpp>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename label_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<label_t>>,
+           std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                                    std::optional<rmm::device_uvector<label_t>>>>>
+prepare_next_frontier(
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> sampled_src_vertices,
+  std::optional<raft::device_span<label_t const>> sampled_src_vertex_labels,
+  raft::device_span<vertex_t const> sampled_dst_vertices,
+  std::optional<raft::device_span<label_t const>> sampled_dst_vertex_labels,
+  std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                           std::optional<rmm::device_uvector<label_t>>>>&& vertex_used_as_source,
```

Comment:
This includes all the source vertices from the previous hops (excluding the current hop, source vertices from the current hop will be added to this and returned), right? Better add a short comment? I need to read the code to figure this out.

Reply:
Added some documentation in `sampling_utils.hpp`